cmd/start.go

```diff
@@ -698,9 +561,228 @@ func (x *Start) Execute(args []string) error {
 	return nil
 }
 
-func getConfigFiles(datafolder string) error {
+func getIPFSRootHash() (*namepb.IpnsEntry, error) {
+	var err error
+	_, ipnskey := namesys.IpnsKeysForID(ipfsNode.Identity)
+	ival, hasherr := ipfsNode.Repo.Datastore().Get(dshelp.NewKeyFromBinary([]byte(ipnskey)))
+	if hasherr != nil {
+		log.Error("get ipns key:", hasherr)
+		return &namepb.IpnsEntry{}, hasherr
+	}
+	val := ival.([]byte)
+	dhtrec := new(recpb.Record)
+	err = proto.Unmarshal(val, dhtrec)
+	if err != nil {
+		log.Error("unmarshal record", err)
+		return &namepb.IpnsEntry{}, err
+	}
+	e := new(namepb.IpnsEntry)
+	err = proto.Unmarshal(dhtrec.GetValue(), e)
+	if err != nil {
+		log.Error("unmarshal record value", err)
+		return &namepb.IpnsEntry{}, err
+	}
+	return e, nil
+}
+
+func getNodeConfig() *ipfscore.BuildCfg {
+	nodeConfig := &ipfscore.BuildCfg{
+		Repo:   nodeRepo,
+		Online: true,
+		ExtraOpts: map[string]bool{
+			"mplex": true,
+		},
+		DNSResolver: dnsResolver,
+		Routing:     DHTOption,
+	}
+
+	if ipfsConfig.Ipns.UsePersistentCache {
+		nodeConfig.ExtraOpts["ipnsps"] = true
+	}
+
+	if onionTransport != nil {
+		nodeConfig.Host = defaultHostOption
+	}
+	return nodeConfig
+}
+
+func defaultHostOption(ctx context.Context, id peer.ID, ps pstore.Peerstore, bwr metrics.Reporter, fs []*net.IPNet, tpt smux.Transport, protec ipnet.Protector, opts *ipfscore.ConstructPeerHostOpts) (p2phost.Host, error) {
+	// no addresses to begin with. we'll start later.
+	swrm, err := swarm.NewSwarmWithProtector(ctx, nil, id, ps, protec, tpt, bwr)
+	if err != nil {
+		return nil, err
+	}
+
+	network := (*swarm.Network)(swrm)
+	network.Swarm().AddTransport(onionTransport)
+
+	for _, f := range fs {
+		network.Swarm().Filters.AddDialFilter(f)
+	}
+
+	var host *p2pbhost.BasicHost
+	if usingTor && !usingClearnet {
+		host = p2pbhost.New(network)
+	} else {
+		hostOpts := []interface{}{bwr}
+		if !opts.DisableNatPortMap {
+			hostOpts = append(hostOpts, p2pbhost.NATPortMap)
+		}
+		host = p2pbhost.New(network, hostOpts...)
+	}
+	return host, nil
+}
+
+func setupTorTransport(password string) error {
+	var err error
+	torControl := torConfig.TorControl
+	if torControl == "" {
+		controlPort, err = obnet.GetTorControlPort()
+		if err != nil {
+			log.Error("get tor control port:", err)
+			return err
+		}
+		torControl = "127.0.0.1:" + strconv.Itoa(controlPort)
+	}
+	torPw := torConfig.Password
+	if password != "" {
+		torPw = password
+	}
+	onionTransport, err = oniontp.NewOnionTransport("tcp4", torControl, torPw, nil, repoPath, (usingTor && usingClearnet))
+	if err != nil {
+		log.Error("setup tor transport:", err)
+		return err
+	}
+	return nil
+}
```

Comment:
This method mutates the node and would better to pass in the mutated objects as args instead of pulling them out of the global vars. Limits mutation to this function and will make it testable later.